### DISCUSSION TITLE
ModbusServerETH.h

### DISCUSSION THREAD

**Discussion by mauricemullens** [eModbus/eModbus#357] [General]
Hi,

when i include ModbusServerETH.h and use ModbusServerEthernet MBserver i get this error:
ModbusServerTCPtemp.h:224:8: error: cannot declare variable 'server' to be of abstract type 'WiFiServer'

i am using an WT32ETH-01 board, and have a wired connection setup. I am also using the async client, which works perfectly. 

Any idea what is wrong?

Thanks
Maurice

**Comment by Miq1** [comment]
Can you try this short code please?
```
#include <Arduino.h>
#include <ModbusServerETH.h>

ModbusServerEthernet MSE;

ModbusMessage FC03(ModbusMessage request) {
  return ECHO_RESPONSE;
}

void setup() {
  Serial.begin(115200);
  Serial.print  (__DATE__);
  Serial.print  (" ");
  Serial.println(__TIME__);

  int result = ETH.begin(ETH_PHY_ADDR, ETH_PHY_POWER);
  if (result) {
    delay(2000);
    ETH.setHostname("WT32-01");

    Serial.print("result: "); Serial.println(result);
    Serial.print("ETH MAC: "); Serial.println(ETH.macAddress());
    Serial.print("IPv4: "); Serial.println(ETH.localIP());
    Serial.print("fullDuplex: "); Serial.println(ETH.fullDuplex());
    Serial.print("linkSpeed: "); Serial.println(ETH.linkSpeed());

    MSE.registerWorker(1, READ_HOLD_REGISTER, FC03);
    MSE.start(502, 4, 2000);
  } else {
    Serial.println("ETH.begin failed...");
    while (true) {
      delay(1000);
    }
  }
}

void loop() {
}
```

**Reply by mauricemullens** [reply]
i did, same error:
eModbus/src/ModbusServerTCPtemp.h:224:8: error: cannot declare variable 'server' to be of abstract type 'WiFiServer'
     ST server(myself->serverPort);

**Comment by Miq1** [comment]
What core version are you using? These are my versions:
```
PACKAGES: 
 - framework-arduinoespressif32 @ 3.20014.231204 (2.0.14) 
 - tool-esptoolpy @ 1.40501.0 (4.5.1) 
 - toolchain-xtensa-esp32 @ 8.4.0+2021r2-patch5
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 36 compatible libraries
Scanning dependencies...
Dependency Graph
|-- eModbus @ 1.7.2
|-- Ethernet @ 2.0.0
|-- SPI @ 2.0.0
```
I only recently switched to the standard Ethernet library after some years of using a fork by @maxgerhardt. If you still have that you may run into trouble.

**Reply by mauricemullens** [reply]
[env:wt32-eth01]
platform = espressif32
board = wt32-eth01
framework = arduino
monitor_speed = 115200
upload_protocol = esptool
lib_deps = 
	SPI@2.0.0
	Ethernet@2.0.0
	; chrisjoyce911/esp32FOTA@^0.2.7
	miq19/eModbus@^1.7.2
	; esphome/ESPAsyncWebServer-esphome@^3.1.0
	; hwspeedy/arduino-mqtt@^2.5.1

**Comment by Miq1** [comment]
My ``platformio.ini`` is even simpler:
```
[env:wt32-eth01]
platform = espressif32
board = wt32-eth01
framework = arduino
monitor_speed = 115200
lib_deps = eModbus
build_flags = -DLOG_LEVEL=6
```
But it looks like the dependency checker has picked the same versions.

If you right click on the ``WiFiServer`` item in ``ModbusServerETH.h`` and chose "Go to Definition" - where do you end up?
Here it is
![grafik](https://github.com/eModbus/eModbus/assets/25884124/0674ba75-8415-43fc-8603-8f1a0f9046da)

**Comment by mauricemullens** [comment]
i end up at: WiFiServer.h

*
  Server.h - Server class for Raspberry Pi
  Copyright (c) 2016 Hristo Gochkov  All right reserved.

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/
#ifndef _WIFISERVER_H_
#define _WIFISERVER_H_

#include "Arduino.h"
#include "Server.h"
#include "WiFiClient.h"
#include "IPAddress.h"

class WiFiServer : public Server {
  private:
    int sockfd;
    int _accepted_sockfd = -1;
    IPAddress _addr;
    uint16_t _port;
    uint8_t _max_clients;
    bool _listening;
    bool _noDelay = false;

**Reply by mauricemullens** [reply]
<img width="623" alt="image" src="https://github.com/eModbus/eModbus/assets/16374915/349485fc-944a-44eb-a04b-5ae3961dbde2">

**Reply by Miq1** [reply]
Looks good, but the essential question is: is there a ``virtual`` function in the definition of ``WiFiServer``? If so, you somehow have got a ``WiFi.h`` different from the regular in your core code.

**Reply by mauricemullens** [reply]
i cant see any virtual functions.  Checked multiple files. Could it be a platformio version issue? I am going to check on another pc, and try it clean from there.

**Reply by Miq1** [reply]
So it looks like the compiler takes a different core than VSC/Intellisense. You may habe to check the xtensa include setup and compare it against the VSC paths. 

If all else fails you may be better off renaming your vscode and platformio folders and start from scratch - installing PlatformIO again, set up a fresh project and give that a try.

**Comment by mauricemullens** [comment]
I just tested it on my Windows pc and everything compiled just fine. So, on my mac, there indeed seems to be a problem with the compiler versions. I will just do a complete reinstall and see if the problem persists on OSX. I will keep you informed. Thanks very much for all the assistance
